### ISSUE TITLE
traceroute parser unable to grab IP when using "-n"

### ISSUE THREAD

**Issue by lelutin** [kellyjonbrazil/jc#272]
Hi! I'm using jc's traceroute parser to automate networking tests with pytest. it's pretty damn useful!

I'm using jc version 1.14.3 on debian bullseye for the tests, but I've also verified that the same behaviour is happening with jc 1.20.0 in debian sid

I've just hit a tiny snag while using it: apparently when we give traceroute a `-n` argument, the parser is not able to grab the IP adresses anymore.

I'm wondering if the parser needs to match the hostname only conditionally.. but I have not yet taken a look at the parser's code.

Here's an example of what happens -- I've added `-q 1` just to make the output more concise but the same happens without it (the number of hops seems correct but all of the "ip" fields are empty):

~~~
$ jc -p traceroute -n -q1 www.koumbit.org
{
  "destination_ip": "199.58.80.40",
  "destination_name": "www.koumbit.org",
  "hops": [
    {
      "hop": 1,
      "probes": [
        {
          "annotation": null,
          "asn": null,
          "ip": null,
          "name": null,
          "rtt": 0.95
        }
      ]
    },
    {
      "hop": 2,
      "probes": [
        {
          "annotation": null,
          "asn": null,
          "ip": null,
          "name": null,
          "rtt": 12.635
        }
      ]
    },
    {
      "hop": 3,
      "probes": [
        {
          "annotation": null,
          "asn": null,
          "ip": null,
          "name": null,
          "rtt": 20.982
        }
      ]
    },
    {
      "hop": 4,
      "probes": [
        {
          "annotation": null,
          "asn": null,
          "ip": null,
          "name": null,
          "rtt": 19.329
        }
      ]
    },
    {
      "hop": 5,
      "probes": [
        {
          "annotation": null,
          "asn": null,
          "ip": null,
          "name": null,
          "rtt": 20.229
        }
      ]
    },
    {
      "hop": 6,
      "probes": [
        {
          "annotation": null,
          "asn": null,
          "ip": null,
          "name": null,
          "rtt": 20.887
        }
      ]
    },
    {
      "hop": 7,
      "probes": [
        {
          "annotation": null,
          "asn": null,
          "ip": null,
          "name": null,
          "rtt": 20.315
        }
      ]
    },
    {
      "hop": 8,
      "probes": [
        {
          "annotation": null,
          "asn": null,
          "ip": null,
          "name": null,
          "rtt": 21.653
        }
      ]
    }
  ]
}
~~~

**Comment by kellyjonbrazil** [comment]
Thank you for reporting this! I’ll take a look - hopefully this is a simple regex fix. Could you give me a couple examples of the traceroute output? That will make it quicker to fix. Thanks!

**Comment by lelutin** [comment]
sure! the same traceroute command without jc gives me:

~~~
$ traceroute -n -q1 www.koumbit.org
traceroute to www.koumbit.org (199.58.80.40), 30 hops max, 60 byte packets
 1  192.168.2.1  3.425 ms
 2  24.212.242.17  16.153 ms
 3  10.170.192.58  17.231 ms
 4  192.171.61.161  25.393 ms
 5  206.248.189.97  25.322 ms
 6  198.179.18.41  23.755 ms
 7  64.15.69.54  25.091 ms
 8  199.58.80.40  25.196 ms
~~~

and without the `-q 1` option:

~~~
$ traceroute -n www.koumbit.org
traceroute to www.koumbit.org (199.58.80.40), 30 hops max, 60 byte packets
 1  192.168.2.1  0.967 ms  1.022 ms  1.204 ms
 2  24.212.242.17  10.176 ms  18.136 ms  18.244 ms
 3  10.170.192.58  19.396 ms  19.575 ms  19.572 ms
 4  192.171.61.61  23.072 ms 206.248.155.109  17.073 ms 192.171.63.17  23.308 ms
 5  206.248.189.97  20.521 ms  22.837 ms  23.194 ms
 6  198.179.18.41  18.334 ms  17.894 ms  17.792 ms
 7  64.15.69.54  17.056 ms  14.033 ms  12.351 ms
 8  199.58.80.40  18.203 ms  18.789 ms  18.906 ms
~~~

**Comment by kellyjonbrazil** [comment]
I added `-n` support in the `dev` branch. Could you copy the traceroute.py file to your `jc` plugin directory and test? (note, I have not tested w/IPv6 yet)

https://github.com/kellyjonbrazil/jc/blob/dev/jc/parsers/traceroute.py

Here are the plugin directories:

- Linux/unix: `$HOME/.local/share/jc/jcparsers`
- macOS: `$HOME/Library/Application Support/jc/jcparsers`
- Windows: `$LOCALAPPDATA\jc\jc\jcparsers`

**Comment by lelutin** [comment]
wow that was fast! I can confirm that the new version of traceroute.py does still work without `-n` and now grabs IPv4 addresses with `-n` given to traceroute.

I've also tested it on a host with IPv6 and it seems that we'll need an additional fix for that address family too:

~~~
$ jc traceroute -n -6 -q1 www.google.com
{"destination_ip":"2607:f8b0:4020:806::2004","destination_name":"www.google.com","hops":[{"hop":1,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt":4.685}]},{"hop":2,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt":0.365}]},{"hop":3,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt":0.303}]},{"hop":4,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt":18.179}]},{"hop":5,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt:12.739}]},{"hop":6,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt":10.113}]},{"hop":7,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt":8.47}]},{"hop":8,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt":9.257}]},{"hop":9,"probes":[]},{"hop":10,"probes":[]},{"hop":11,"probes":[{"annotation":null,"asn":null,"ip":null,"name":null,"rtt":8.66}]}]}
~~~

without -n, works ok:

~~~
$ jc traceroute -6 -q1 www.google.com
{"destination_ip":"2607:f8b0:4020:806::2004","destination_name":"www.google.com","hops":[{"hop":1,"probes":[{"annotation":null,"asn":null,"ip":"2605:9000:402:6a01::1","name":"2605:9000:402:6a01::1","rtt":3.875}]},{"hop":2,"probes":[{"annotation":null,"asn":null,"ip":"2605:9000:0:400a::f1","name":"2605:9000:0:400a::f1","rtt":0.377}]},{"hop":3,"probes":[{"annotation":null,"asn":null,"ip":"2605:9000:0:101::1","name":"2605:9000:0:101::1","rtt":1.012}]},{"hop":4,"probes":[{"annotation":null,"asn":null,"ip":"2001:5a0:300:200::202","name":"if-ae-0-2.tcore2.mtt-montreal.ipv6.as6453.net","rtt":12.723}]},{"hop":5,"probes":[{"annotation":null,"asn":null,"ip":"2001:5a0:300:200::202","name":"if-ae-0-2.tcore2.mtt-montreal.ipv6.as6453.net","rtt":14.416}]},{"hop":6,"probes":[{"annotation":null,"asn":null,"ip":"2001:5a0:1900:100::12","name":"if-ae-5-2.tcore2.n0v-newyork.ipv6.as6453.net","rtt":10.378}]},{"hop":7,"probes":[{"annotation":null,"asn":null,"ip":"2001:4860:0:1127::14","name":"2001:4860:0:1127::14","rtt":9.289}]},{"hop":8,"probes":[{"annotation":null,"asn":null,"ip":"2001:4860::c:4002:652a","name":"2001:4860::c:4002:652a","rtt":8.946}]},{"hop":9,"probes":[{"annotation":null,"asn:null,"ip":"2001:4860::c:4002:6523","name":"2001:4860::c:4002:6523","rtt":9.456}]},{"hop":10,"probes":[{"annotation":null,"asn":null,"ip":"2001:4860:0:1::c73","name":"2001:4860:0:1::c73","rtt":9.63}]},{"hop":11,"probes":[{"annotation":null,"asn":null,"ip":"2001:4860:0:11da::1","name":"2001:4860:0:11da::1","rtt":8.618}]},{"hop":12,"probes":[{"annotation":null,"asn":null,"ip":"2001:4860:0:1::d71","name":"2001:4860:0:1::d71","rtt":9.284}]},{"hop":13,"probes":[{"annotation":null,"asn":null,"ip":"2607:f8b0:4020:806::2004","name":"yul03s04-in-x04.1e100.net","rtt":8.223}]}]}
~~~

as before, here's an example of the ipv6 traceroute with -n set:

~~~
$ traceroute -n -6 www.google.com
traceroute to www.google.com (2607:f8b0:4020:806::2004), 30 hops max, 80 byte packets
 1  2605:9000:402:6a01::1  4.181 ms  4.294 ms  4.253 ms
 2  2605:9000:0:400a::f1  0.354 ms  0.532 ms  0.484 ms
 3  2001:5a0:40:100::51  15.284 ms 2605:9000:0:101::1  4.864 ms 2001:5a0:40:100::51  15.415 ms
 4  2001:5a0:40:100::51  15.379 ms 2001:5a0:300:200::202  12.709 ms 2001:5a0:40:100::51  15.289 ms
 5  2001:5a0:1900:100::12  10.020 ms  10.212 ms  10.163 ms
 6  2001:5a0:1900:100::12  10.113 ms 2001:5a0:400:700::17  8.399 ms 2001:5a0:1900:100::12  10.215 ms
 7  2001:4860:0:1127::2  9.110 ms 2001:5a0:400:700::17  8.476 ms  8.380 ms
 8  2001:4860::8:4000:cd80  9.428 ms 2001:4860:0:1128::14  9.360 ms 2001:4860::8:4000:cd80  9.229 ms
 9  2001:4860::9:4001:d508  9.376 ms 2001:4860::c:4002:652a  9.105 ms 2001:4860::c:4002:6523  9.384 ms
10  2001:4860:0:11da::1  8.489 ms 2001:4860::9:4001:d508  8.978 ms 2001:4860::1c:4000:f5eb  9.640 ms
11  2001:4860:0:1::c73  9.596 ms  9.077 ms  9.724 ms
12  2607:f8b0:4020:806::2004  8.086 ms  8.091 ms  8.436 ms
~~~

and without -n for comparison:

~~~
$ traceroute -6 www.google.com
traceroute to www.google.com (2607:f8b0:4020:806::2004), 30 hops max, 80 byte packets
 1  2605:9000:402:6a01::1 (2605:9000:402:6a01::1)  4.630 ms  4.748 ms  4.642 ms
 2  2605:9000:0:400a::f1 (2605:9000:0:400a::f1)  0.381 ms 2605:9000:0:400b::f2 (2605:9000:0:400b::f2)  0.325 ms  0.456 ms
 3  2605:9000:0:101::1 (2605:9000:0:101::1)  0.362 ms * *
 4  if-ae-0-2.tcore2.mtt-montreal.ipv6.as6453.net (2001:5a0:300:200::202)  12.898 ms  16.707 ms ix-xe-1-1-0-0.tcore1.mtt-montreal.ipv6.as6453.net (2001:5a0:40:100::51)  13.266 ms
 5  if-ae-0-2.tcore2.mtt-montreal.ipv6.as6453.net (2001:5a0:300:200::202)  13.601 ms  16.389 ms if-ae-5-2.tcore2.n0v-newyork.ipv6.as6453.net (2001:5a0:1900:100::12)  10.247 ms
 6  if-ae-5-2.tcore2.n0v-newyork.ipv6.as6453.net (2001:5a0:1900:100::12)  10.422 ms *  10.412 ms
 7  2001:5a0:400:700::17 (2001:5a0:400:700::17)  8.669 ms 2001:4860:0:1128::3 (2001:4860:0:1128::3)  9.236 ms *
 8  2001:4860::c:4002:6523 (2001:4860::c:4002:6523)  9.426 ms 2001:4860:0:1128::14 (2001:4860:0:1128::14)  9.603 ms 2001:4860:0:1127::2 (2001:4860:0:1127::2)  9.202 ms
 9  2001:4860::9:4001:d508 (2001:4860::9:4001:d508)  9.663 ms 2001:4860::c:4002:6523 (2001:4860::c:4002:6523)  9.358 ms 2001:4860::9:4001:d508 (2001:4860::9:4001:d508)  9.167 ms
10  2001:4860::9:4000:f9d9 (2001:4860::9:4000:f9d9)  9.788 ms 2001:4860::1c:4000:f5eb (2001:4860::1c:4000:f5eb)  9.852 ms 2001:4860:0:11da::1 (2001:4860:0:11da::1)  8.633 ms
11  2001:4860:0:1::d71 (2001:4860:0:1::d71)  9.641 ms yul03s04-in-x04.1e100.net (2607:f8b0:4020:806::2004)  8.244 ms 2001:4860:0:1::d71 (2001:4860:0:1::d71)  9.526 ms
~~~

**Comment by kellyjonbrazil** [comment]
Thanks! isolating IPv6 addresses will be a little trickier, but I'll work on that.

**Comment by kellyjonbrazil** [comment]
Ok, I think I have IPv6 working now, too. If you could try the latest version of that same file that would be great. Thanks!

**Comment by lelutin** [comment]
nice! I've tested again the 4 cases with the new version. ipv4 with(out) `-n` and ipv6 with(out) `-n` and it all seems to be working great. many thanks for such a quick fix! :)

**Comment by kellyjonbrazil** [comment]
Excellent! I'll add this fix to the next release in a few days. Thanks so much for the testing and feedback.

**Comment by kellyjonbrazil** [comment]
Released in v1.21.0 available via pip. Binaries coming soon.
